internal/cmd/agent/deployer/driftdetect/driftdetect.go

```diff
@@ -1,57 +1,147 @@
 package driftdetect
 
 import (
+	"context"
+	"math/rand"
+
+	"github.com/go-logr/logr"
+
 	"github.com/rancher/fleet/internal/cmd/agent/deployer/plan"
+	"github.com/rancher/fleet/internal/cmd/agent/trigger"
 	"github.com/rancher/fleet/internal/helmdeployer"
 	fleet "github.com/rancher/fleet/pkg/apis/fleet.cattle.io/v1alpha1"
 
 	"github.com/rancher/wrangler/v2/pkg/apply"
 
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/client-go/util/retry"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
 type DriftDetect struct {
-	defaultNamespace string
-	deployer         *helmdeployer.Helm
+	// Trigger watches deployed resources on the local cluster.
+	trigger *trigger.Trigger
+
+	upstreamClient client.Client
+	upstreamReader client.Reader
+
 	apply            apply.Apply
+	defaultNamespace string
 	labelPrefix      string
 	labelSuffix      string
 }
 
-func New(defaultNamespace string,
-	labelPrefix, labelSuffix string,
-	deployer *helmdeployer.Helm,
-	apply apply.Apply) *DriftDetect {
+func New(
+	trigger *trigger.Trigger,
+	upstreamClient client.Client,
+	upstreamReader client.Reader,
+	apply apply.Apply,
+	defaultNamespace string,
+	labelPrefix string,
+	labelSuffix string,
+) *DriftDetect {
 	return &DriftDetect{
+		trigger:          trigger,
+		upstreamClient:   upstreamClient,
+		upstreamReader:   upstreamReader,
+		apply:            apply.WithDynamicLookup(),
 		defaultNamespace: defaultNamespace,
 		labelPrefix:      labelPrefix,
 		labelSuffix:      labelSuffix,
-		deployer:         deployer,
-		apply:            apply.WithDynamicLookup(),
 	}
 }
 
-// AllResources returns the resources that are deployed by the bundle deployment,
-// according to the helm release history. It adds to be deleted resources to
-// the list, by comparing the desired state to the actual state with apply.
-func (m *DriftDetect) AllResources(bd *fleet.BundleDeployment) (*helmdeployer.Resources, error) {
-	resources, err := m.deployer.Resources(bd.Name, bd.Status.Release)
+func (d *DriftDetect) Clear(bdKey string) error {
+	return d.trigger.Clear(bdKey)
+}
+
+func (d *DriftDetect) Refresh(logger logr.Logger, bdKey string, bd *fleet.BundleDeployment, resources *helmdeployer.Resources) (*fleet.BundleDeployment, error) {
+	logger = logger.WithName("DriftDetect")
+	if bd == nil {
+		return bd, d.trigger.Clear(bdKey)
```

Comment:
This will never be reached as bd will never be nil when this is called. Again finalizers can help here. Happy to discuss this

Reply:
We should call trigger.Clear, currently we are not doing it.  We are leaving the watches forever